compiler/il/Aliases.cpp

```diff
@@ -542,8 +542,11 @@ OMR::SymbolReference::getUseDefAliasesBV(bool isDirectCall, bool includeGCSafePo
             return aliases;
             }
 #endif
-
-         return symRefTab->aliasBuilder.methodAliases(self());
+         TR_BitVector * aliases = new (aliasRegion) TR_BitVector(bvInitialSize, aliasRegion, growability);
+         *aliases |= symRefTab->aliasBuilder.addressTakenAutos();
+         TR_BitVector *methodAliases = symRefTab->aliasBuilder.methodAliases(self());
```

Comment:
I wonder if you could check if addressTakenAutos is non empty before create the new "aliases" bit vector in this case. i.e. if it was empty, then we could just return the same answer as in the original code without creating a new bit vector (especially for front ends that will never have bits set in addressTakenAutos).

Reply:
Yes makes sense. I will also create a test case before resubmitting.

Reply:
BTW I am getting an out of bounds error again if I run with more optimizations ... so there is probably something missing that I need to add.

[trtrace.log](https://github.com/eclipse/omr/files/2280363/trtrace.log)

Reply:
Which optimization causes your error ? It may be that some optimization passes need some (hopefully small) change to account for the fact that autos could be in the alias sets returned for calls. This kind of aliasing may not have happened before in the front ends that are running the optimizations that you are running.

Reply:
Okay I see you were doing global copy propagation when the crash happened. Which IL node was that analysis looking at and what was the stack trace for the crash ?

Reply:
@dibyendumajumdar just a gentle reminder there's an outstanding question for you here...

Reply:
Hi - yes, sorry just been busy with other things. I will do some more tests and report back.

Reply:
@dibyendumajumdar just a friendly ping to check in whether you had some more information on the crash (the earlier question) so that we can try to help. Thanks.

Reply:
@vijaysun-omr yes sorry for the delay. I have finally got some time to look at this. Creating test case and then will update this pull request with more info. Thanks

Reply:
@vijaysun-omr Stack trace:
```
>	sparse-omrjit.exe!std::deque<unsigned int,TR::typed_allocator<unsigned int,TR::Region &> >::operator[](unsigned __int64 _Pos) Line 1411	C++
 	sparse-omrjit.exe!TR::deque<unsigned int,TR::Region &>::operator[](unsigned __int64 index) Line 108	C++
 	sparse-omrjit.exe!TR_UseDefInfo::insertData(TR::Block * block, TR::Node * node, TR::Node * parent, TR::TreeTop * treeTop, TR_UseDefInfo::AuxiliaryData & aux, CS2::ASparseBitVector<CS2::shared_allocator<CS2::heap_allocator<65536,12,TRMemoryAllocator<1,12,28> > > > & expandedLIndexes, bool considerImplicitStores) Line 1660	C++
 	sparse-omrjit.exe!TR_UseDefInfo::insertData(TR::Block * block, TR::Node * node, TR::Node * parent, TR::TreeTop * treeTop, TR_UseDefInfo::AuxiliaryData & aux, CS2::ASparseBitVector<CS2::shared_allocator<CS2::heap_allocator<65536,12,TRMemoryAllocator<1,12,28> > > > & expandedLIndexes, bool considerImplicitStores) Line 1559	C++
 	sparse-omrjit.exe!TR_UseDefInfo::fillInDataStructures(TR_UseDefInfo::AuxiliaryData & aux) Line 485	C++
 	sparse-omrjit.exe!TR_UseDefInfo::prepareUseDefInfo(bool requiresGlobals, bool prefersGlobals, bool cannotOmitTrivialDefs, bool conversionRegsOnly) Line 323	C++
 	sparse-omrjit.exe!TR_UseDefInfo::TR_UseDefInfo(TR::Compilation * comp, TR::CFG * cfg, TR::Optimizer * optimizer, bool requiresGlobals, bool prefersGlobals, bool loadsShouldBeDefs, bool cannotOmitTrivialDefs, bool conversionRegsOnly, bool doCompletion) Line 103	C++
 	sparse-omrjit.exe!OMR::Optimizer::performOptimization(const OptimizationStrategy * optimization, int firstOptIndex, int lastOptIndex, int doTiming) Line 1850	C++
 	sparse-omrjit.exe!OMR::Optimizer::optimize() Line 1129	C++
 	sparse-omrjit.exe!OMR::Compilation::performOptimizations() Line 1275	C++
 	sparse-omrjit.exe!OMR::Compilation::compile() Line 1058	C++
 	sparse-omrjit.exe!compileMethodFromDetails(OMR_VMThread * omrVMThread, TR::IlGeneratorMethodDetails & details, TR_Hotness hotness, int & rc) Line 373	C++
 	sparse-omrjit.exe!nj::FunctionBuilder::compile(int opt_level) Line 261	C++
 	sparse-omrjit.exe!JIT_Compile(JIT_FunctionBuilder * fb, int opt_level) Line 451	C++
 	sparse-omrjit.exe!output_fn(dmr_C * C, JIT_Context * module, entrypoint * ep) Line 2175	C
 	sparse-omrjit.exe!compile(dmr_C * C, JIT_Context * module, symbol_list * list) Line 2210	C
 	sparse-omrjit.exe!dmrC_omrcompile(int argc, char * * argv, JIT_Context * module, const char * inputbuffer) Line 2249	C
 	sparse-omrjit.exe!main(int argc, char * * argv) Line 90	C
 	[External Code]	
```

Reply:
Node is icall:

```

  | Name | Value | Type
-- | -- | -- | --
◢ | node | 0x000001e54b593bb0 {...} | TR::Node *
  | ◢ OMR::Node | {_opCode={...} _numChildren=2 _visitCount=17 ...} | OMR::Node
  | ◢ _opCode | {...} | TR::ILOpCode
  | ▶ OMR::ILOpCode | {_opCode=icall (61) } | OMR::ILOpCode
  | _numChildren | 2 | unsigned short
  | _visitCount | 17 | unsigned short
  | _globalIndex | 17 | unsigned int
  | ▶ _flags | {_flags=0 } | OMR::FlagsTemplate<unsigned int>
  | _localIndex | 4 | unsigned int
  | ▶ _byteCodeInfo | {_doNotProfile=0 _isSameReceiver=0 _callerIndex=-1 ...} | TR_ByteCodeInfo
  | _referenceCount | 1 | unsigned int
  | ▶ _unionA | {_useDefIndex=3 _register=0x0000000000000003 {...} } | OMR::Node::UnionA
  | ▶ _unionBase | {_constValue=2084323277616 _fpConstValue=14236464.0 _dpConstValue=1.029792526297e-311#DEN ...} | OMR::Node::UnionBase
  | ▶ _unionPropertyA | {_symbolReference=0x000001e54b59c810 {...} _regLoadStoreSymbolReference=0x000001e54b59c810 {...} _branchDestinationNode=...} | OMR::Node::UnionPropertyA

```

Reply:
aux._sideTableToUseDefMap has size 4, so presumably this is why there is an out of bounds exception.

Reply:
So I reckon the bug must be in getExpandedTotalNode()?

 aux._sideTableToUseDefMap.resize(getExpandedTotalNodes());

Reply:
Thanks, could you please upload a "current" version of the log with traceUseDefs added to your command line options ? It was'nt clear to me if stuff had changed in your build since the last log was generated (?)

Reply:
Calls "expand" into N bits in the use def info analysis, one for each symbol reference that may be aliased to the call. This can probably be explained most easily with an example:

call foo // aliases t1, t2 and t3 (say)
...
t1 = rhs;
...
load t1 // the def is the above store 
...
load t2 // the def is the above call
...
load t3 // the def is the above call

So we need a mechanism that kills the call as a def for t1 alone when the store to t1 is encountered and this is what the expanded index achieves...by giving one bit per sym ref that may alias the call, there are separate bits for t1, t2, t3 associated with the call. 

Normally the set of symbol references that may be aliased to the call only contains shadows and statics. So, we may need to check how many bits the call is expanding out to and which aliases are being taken into account in that expansion. In particular are the new address taken autos included in the set of symbol reference that may be aliased when deciding how many bits the call's aliasing should expand to.

Reply:
I have saved it here:
https://github.com/dibyendumajumdar/dmr_c/blob/master/tests/omrjit/traces/trtrace_traceUseDefs.log

I will add another one with full all other options added

Reply:
[trtrace.log](https://github.com/eclipse/omr/files/2441191/trtrace.log)

This is trace with TR_Options=traceUseDefs,log=trtrace.log,traceIlGen,traceFull,traceAliases

Reply:
Thanks, are you constructing usedef info passing true to requiresGlobals and prefersGlobals ?
i.e. in
TR_UseDefInfo::TR_UseDefInfo(TR::Compilation *comp, TR::CFG *cfg, TR::Optimizer *optimizer,
bool requiresGlobals, bool prefersGlobals, bool loadsShouldBeDefs, bool cannotOmitTrivialDefs, bool conversionRegsOnly, bool doCompletion)

I think we may not even be indexing calls as participants in usedef info analysis until those flags are set. I'm still not certain it will all work with just those flags set but I'd be interested in seeing a fresh log and info about any crash that you run into.

Reply:
Hi @vijaysun-omr , it seems that requiresGlobals was set to true, and prefersGlobals set to false. So I changed latter to true too, and attached is the trace log after that.

[trtrace.log](https://github.com/eclipse/omr/files/2454183/trtrace.log)

I would like to understand the aliasing code and fix it for this scenario, but it isn't very clear how to get started. I don't tend to have a lot of time to delve into it so that doesn't help. If you could point me to bit where calls are analyzed then maybe I can start there.

Reply:
@dibyendumajumdar I understand and I will try to help as much as possible to allow you to move forward. The new trace log does not really show the call node as participating in the use def info analysis (i.e. it did not change anything). 

The aliasing info shown in your log file may not necessarily be the problem; from my quick scan it looked fine to me. The issue may be the usage of aliasing info in the specific analysis you are crashing in, namely use def info analysis. The call node that you have in your test case has to participate in the use def info analysis info for it's aliasing info to be even consulted. 

The routine inside use def info analysis that I feel is'nt picking up call nodes is:

bool TR_UseDefInfo::indexSymbolsAndNodes(AuxiliaryData &aux)

This routine is responsible for figuring out which nodes need to participate in the analysis and which index to assign each node and symbol. Not assigning a node an index means that it does not participate in the analysis in any way, i.e. it is considered neither a use nor a definition of anything. 

As an example from the Java usage in OpenJ9, one would expect load and store nodes for auto symbols to participate in the analysis but given that we don't run use def info analysis for globals (i.e. shadows and statics) in OpenJ9, no other kinds of symbols participate in the analysis. The reason why we don't do use def analysis for "globals", i.e. statics and shadows in OpenJ9 is simply a choice that we have made given that aliasing tends to be quite conservative for calls (that can more or less kill any global) and so we don't really anticipate the info to lead to a significant enough throughput performance gain to be worth the compile time cost.

For other OMR-based language compilers that want to track uses and definitions of "global" symbols as well, i.e. statics and shadows, the first step would be to set those two flags I mentioned in the use def info as you are doing already it seems. In that case I would have expected you to index your call node around the following line in UseDefInfo.cpp:

      if (sym->isMethod())
               {
               if (!_hasCallsAsUses)
                  continue;
               }
            else if (!isValidAutoOrParm(symRef))
               continue;

            // Index this symbol
            //
            if (trace())
               {
                  traceMsg(comp(), "Non Trivial symbol [%p] has index %d\n", sym, _numSymbols);
               }
            _sideTableToSymRefNumMap[_numSymbols] = symRefNumber;
            sym->setLocalIndex(_numSymbols++);
_numNonTrivialSymbols++;

We have had other compilers use the "globals" implementation of use def info in the past but because it's not used in OpenJ9, I'll probably have to do some of the discovery of why this mode of running use def info is'nt working for you, i.e. it's not a path I'm as familiar with. That having been said, I am pretty sure we can work together to find the root cause and it probably won't be that hard to fix.

Reply:
Hi @vijaysun-omr, thank you for the detailed answer. I will spend some time on this over the coming weekend (hopefully!) and report back on progress.

Reply:
Also, the following code is where the call node ought to be indexed. So you can check what we did if/when we reached here for your call now.

 else if (opCode.isCall() || opCode.isFence())
         {
         if (num_aliases > 0)
            {
            if (symRefToLocalIndexMap[symRef->getReferenceNumber()] == 0)
               {
               localIndex = _numExpandedDefUseNodes;
               _numExpandedDefUseNodes += num_aliases;
               //traceMsg(comp(), "num_aliases #%d  %d %d %d\n", node->getSymbolReference()->getReferenceNumber(), num_aliases, opCode.isFence(), node->getSymbolReference()->getSymbol()->isMethod());
               useDefIndex = _numDefUseNodes++;
               //             traceMsg(comp(), "UDI: setting useDefIndex to %d _numDefUseNodes = %d\n",useDefIndex,_numDefUseNodes);

               symRefToLocalIndexMap[symRef->getReferenceNumber()] = localIndex;
               }
            else
               {
               localIndex = symRefToLocalIndexMap[symRef->getReferenceNumber()];
               useDefIndex = _numDefUseNodes++;
               }
}

Reply:
> In that case I would have expected you to index your call symbol around the following line in UseDefInfo.cpp:
> 
> ```
>   if (sym->isMethod())
>            {
>            if (!_hasCallsAsUses)
>               continue;
>            }
>         else if (!isValidAutoOrParm(symRef))
>            continue;
>

It seems that _hasCallsAsUses is never set anywhere - perhaps some code related to this was removed?
Anyway, so the reason nothing happens here is because of that.
 
>         // Index this symbol
>         //
>         if (trace())
>            {
>               traceMsg(comp(), "Non Trivial symbol [%p] has index %d\n", sym, _numSymbols);
>            }
>         _sideTableToSymRefNumMap[_numSymbols] = symRefNumber;
>         sym->setLocalIndex(_numSymbols++);
> ```
> 
> So the next step would be to understand if/why the call symbol was not indexed at this point since this probably holds you back from indexing the call node later from:

Reply:
So if I set _hasCallsAsUses to true then it doesn't crash. The test passes so presumably nothing bad happens. So what should be setting this to true? Presumably I need a way to test whether the Call node aliased locals and set this to true in that case?

Reply:
@vijaysun-omr @mstoodle @Leonardo2718  Apart from fixing the issues with optimization, I also need to provide an API to set the address taken flag on a local symbol, so that the alias info is updated. Any views on what would be a preferred interface? In my project I provide an api such as:

```
/**
 * Inform the compiler that the automatic variable had its 
 * address taken, hence the variable should be considered as aliased
 * when calling functions etc.
 */
extern void JIT_SetAutoAddressTaken(JIT_ILInjectorRef ilinjector,
	JIT_SymbolRef symbol);
```

I also need a mechanism in Tril to set this for the test case. 

Thanks and Regards

Reply:
@dibyendumajumdar can you please upload a log from the run where you set _hasCallAsUses to true and the test passed ? With all the same tracing options as you did previously.

Reply:
Hi @vijaysun-omr  

As requested: 
[trtrace.log](https://github.com/eclipse/omr/files/2480526/trtrace.log)

Reply:
Looks like the aliasing did get through to use def info as was intended. As I mentioned before, on OpenJ9 we probably did not need to set that boolean to true since locals cannot be aliased with calls in Java. This boolean should really be set in a language-specific way somehow so as to avoid hardcoding it to either true or false (not needed for Java for compile time reasons and not functionally correct for your language implementation).

Reply:
so to answer the first question setting _hasCallAsUses should be handled as part of the constructor if it needs to be configured by different languages. This is the first time we are going to have a language specific override of a flag like this for UseDefInfo. Looking at how it is currently used instances of TR_UseDef_info are allocated in the OMROptimizer.cpp, ValueNumberInfo.cpp and LoadExtensions.cpp. There is some variation between them in how they configure UseDefInfo for globals. I think the best way to handle this is to create a new API on the Optimizer object (eg OMROptimizer.hpp and OMROptimizer.cpp) which all current allocation sites switch to calling - the method will accept as parameters the bools which vary between the different request sites within the compiler. The value for _hasCallAsUses could be either stored as a flag in the optimizer or by overriding the build method to change how UseDefInfo is built. I would prefer the later since it means languages have a common place to configure UseDefInfo according to their needs. This small refactor would ideally be done as a separate commit and pull request ahead of changes introducing the new flag to the constructor. I hope that makes sense - if it doesn't do keep asking questions and I'll try to clarify more on how this would work.

An added bonus of this is that we could improve the UseDefInfo memory management in a follow-on item which would be a another good reason to refactor the optimizer to allocate UseDefInfo through a central location.

Reply:
Hi @andrewcraik 

Thanks for the feedback. I will open a new pull request for the suggested refactoring.

Regards

Reply:
Hi @andrewcraik @vijaysun-omr , I have submitted https://github.com/eclipse/omr/pull/3866. This contains the refactoring suggested by Andrew above.